filters/auth/jwt_validation.go

```diff
@@ -0,0 +1,301 @@
+	"encoding/json"
+	"fmt"
+	"math/big"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/go-co-op/gocron"
+	"github.com/golang-jwt/jwt"
+	"github.com/opentracing/opentracing-go"
+	log "github.com/sirupsen/logrus"
+	"github.com/zalando/skipper/filters"
+)
+
+const (
+	JwtValidationName = "jwtValidation"
+)
+
+type (
+	jwtValidationSpec struct {
+		options TokenintrospectionOptions
+	}
+
+	jwtValidationFilter struct {
+		scheduler       *gocron.Scheduler
+		claims          []string
+		upstreamHeaders map[string]string
+	}
+)
+
+var m sync.Mutex
+var rsakeys map[string]*rsa.PublicKey = make(map[string]*rsa.PublicKey)
+
+func NewJwtValidationWithOptions(o TokenintrospectionOptions) filters.Spec {
+	return &jwtValidationSpec{
+		options: o,
+	}
+}
+
+func NewJwtValidation(timeout time.Duration) filters.Spec {
+	return NewJwtValidationWithOptions(TokenintrospectionOptions{
+		Timeout: timeout,
+		Tracer:  opentracing.NoopTracer{},
+	})
+}
+
+func JwtValidationWithOptions(
+	create func(time.Duration) filters.Spec,
+	o TokenintrospectionOptions,
+) filters.Spec {
+	s := create(o.Timeout)
+	ts, ok := s.(*jwtValidationSpec)
+	if !ok {
+		return s
+	}
+
+	ts.options = o
+	return ts
+}
+
+func (s *jwtValidationSpec) Name() string {
+	return JwtValidationName
+}
+
+func (s *jwtValidationSpec) CreateFilter(args []interface{}) (filters.Filter, error) {
+	sargs, err := getStrings(args)
+	if err != nil {
+		return nil, err
+	}
+	if len(sargs) < 2 {
+		return nil, filters.ErrInvalidFilterParameters
+	}
+
+	issuerURL := sargs[0]
+
+	cfg, err := getOpenIDConfig(issuerURL)
+	if err != nil {
+		return nil, err
+	}
+
+	scheduler := gocron.NewScheduler(time.UTC)
+	scheduler.Every("30m").Do(func() {
+		err = fillPublicKeys(cfg.JwksURI)
+		if err != nil {
+			log.Errorf("Error getting public keys for url: %s", cfg.JwksURI)
+			return
+		}
+	})
+
+	scheduler.StartAsync()
+
+	f := &jwtValidationFilter{
+		scheduler: scheduler,
+	}
+
+	f.claims = strings.Split(sargs[1], " ")
+	if !all(f.claims, cfg.ClaimsSupported) {
+		return nil, fmt.Errorf("%v: %s, supported Claims: %v", errUnsupportedClaimSpecified, strings.Join(f.claims, ","), cfg.ClaimsSupported)
+	}
+
+	// inject additional headers from the access token for upstream applications
+	if len(sargs) > 2 && sargs[2] != "" {
+		f.upstreamHeaders = make(map[string]string)
+
+		for _, header := range strings.Split(sargs[2], " ") {
+			sl := strings.SplitN(header, ":", 2)
+			if len(sl) != 2 || sl[0] == "" || sl[1] == "" {
+				return nil, fmt.Errorf("%w: malformatted filter for upstream headers %s", filters.ErrInvalidFilterParameters, sl)
+			}
+			f.upstreamHeaders[sl[0]] = sl[1]
+		}
+		log.Debugf("Upstream Headers: %v", f.upstreamHeaders)
+	}
+
+	return f, nil
+}
+
+// String prints nicely the jwtValidationFilter configuration based on the
+// configuration and check used.
+func (f *jwtValidationFilter) String() string {
+	return fmt.Sprintf("%s(%s)", JwtValidationName, strings.Join(f.claims, ","))
+}
```

Comment:
I am not sure we have String() for other filters so lets remove if it is not used